File: 10_phase10_multi_object_tracking/10C_multi_camera_3d_multi_object_tracking/perception/perception_3d_pipeline.py
import cv2
import numpy as np

from cv_bridge import CvBridge
from sensor_msgs_py import point_cloud2
from ultralytics import YOLO

from sensor_msgs.msg import PointCloud2

class Perception3DPipeline:

    def __init__(self):
        self.bridge = None
        self.model = None

        self.cameras = None

        self.fx = None
        self.fy = None
        self.cx = None
        self.cy = None

        self.MIN_OBJECT_POINTS = None

        self.EGO_HOOD_POLYGON = np.array([
            [175, 480],
            [240, 390],
            [400, 390],
            [465, 480]
        ], dtype=np.int32)
        
        self.scale_polygon(self.EGO_HOOD_POLYGON, 1.10)

    def validate(self):
        assert self.bridge is not None
        assert self.model is not None
        assert self.cameras is not None

        assert self.fx is not None
        assert self.fy is not None
        assert self.cx is not None
        assert self.cy is not None

    def convert_ros_to_cv(
        self,
        image_msg
    ):

        return self.bridge.imgmsg_to_cv2(
            image_msg,
            desired_encoding="bgr8"
        )


    def convert_ros_to_numpy(
        self,
        lidar_msg
    ):

        points = np.asarray(
            list(
                point_cloud2.read_points(
                    lidar_msg,
                    field_names=("x", "y", "z"),
                    skip_nans=True
                )
            )
        )

        xyz = np.stack(
            [
                points["x"],
                points["y"],
                points["z"]
            ],
            axis=1
        ).astype(np.float32)

        return xyz


    def yolo_detection(
        self,
        front,
        rear,
        left,
        right
    ):

        front, front_result = self.detect_image(front)

        rear, rear_result = self.detect_image(rear)

        left, left_result = self.detect_image(left)

        right, right_result = self.detect_image(right)

        return (

            front,
            rear,
            left,
            right,

            front_result,
            rear_result,
            left_result,
            right_result

        )

    def scale_polygon(self, polygon, scale=1.05):
        center = polygon.mean(axis=0)
        self.EGO_HOOD_POLYGON = np.round((polygon - center) * scale + center).astype(np.int32)


    def remove_hood(self, image):
        image = image.copy()
        cv2.fillPoly(image, [self.EGO_HOOD_POLYGON], (0, 0, 0))
        return image  


    def detect_image(self, image):

        original_image = image.copy()

        inference_image = self.remove_hood(image)

        results = self.model(
            inference_image,
            imgsz=640,
            verbose=False
        )

        annotated = results[0].plot(img=original_image)

        return annotated, results[0]

    def project_lidar(
        self,
        image,
        lidar,
        camera_name
    ):

        points = lidar.copy() # Camera frame
        ego_points = lidar.copy() # Ego frame

        #
        # Camera configuration
        #

        yaw = -np.deg2rad(
            self.cameras[camera_name]["yaw"]
        )

        #
        # Rotate LiDAR into camera frame
        #

        R = np.array(
            [
                [np.cos(yaw), -np.sin(yaw), 0.0],
                [np.sin(yaw),  np.cos(yaw), 0.0],
                [0.0,          0.0,         1.0]
            ],
            dtype=np.float32
        )

        points = points @ R.T

        #
        # LiDAR -> Camera translation
        #

        points[:, 0] -= 1.5

        #
        # CARLA axes -> Camera axes
        #

        X = points[:, 1]
        Y = -points[:, 2]
        Z = points[:, 0]

        #
        # Keep points in front
        #

        mask = Z > 0

        X = X[mask]
        Y = Y[mask]
        Z = Z[mask]

        if len(Z) == 0:
            return image

        #
        # Projection
        #

        u = self.fx * X / Z + self.cx
        v = self.fy * Y / Z + self.cy

        #
        # Keep points inside image
        #

        h, w = image.shape[:2]

        valid = (
            (u >= 0) &
            (u < w) &
            (v >= 0) &
            (v < h)
        )

        u = u[valid].astype(np.int32)
        v = v[valid].astype(np.int32)

        return (
            image,
            u,
            v,
            ego_points[mask][valid]
        )


    def extract_object_clouds(
        self,
        image,
        u,
        v,
        projected_points,
        results
    ):

        if results.masks is None:
            return []

        # masks = results.masks.data.cpu().numpy()

        masks = []

        for mask in results.masks.data.cpu().numpy():

            mask = cv2.resize(
                mask.astype(np.uint8),
                (image.shape[1], image.shape[0]),
                interpolation=cv2.INTER_NEAREST
            )

            masks.append(mask)        

        object_clouds  = []

        for mask in masks:

            class_id = int(results.boxes.cls[len(object_clouds)])
            class_name = self.model.names[class_id]

            cloud = []

            for i, (x, y) in enumerate(zip(u, v)):

                if mask[int(y), int(x)] > 0:

                    cloud.append(
                        {
                            "pixel": (int(x), int(y)),
                            "xyz": projected_points[i]                            
                        }
                    )

            # object_clouds.append(cloud)

            object_clouds.append(
                {
                    "cloud": cloud,
                    "box": results.boxes.xyxy.cpu().numpy()[len(object_clouds)],
                    "class_name": class_name,
                    "score": float(
                        results.boxes.conf[len(object_clouds)].cpu().item()
                    )
                }
            )

        return object_clouds 


    def process_object_clouds_and_distance(
        self,
        image,
        object_clouds,
        camera_name
    ):

        world_objects = []

        for obj in object_clouds:

            if len(obj["cloud"]) < self.MIN_OBJECT_POINTS:
                continue

            image = self.draw_object_cloud(
                image,
                obj["cloud"]
            )

            #
            # Convert LiDAR cluster once
            #

            xyz_points = self.process_object_clouds(
                obj["cloud"]
            )

            #
            # Estimate object attributes
            #

            distance = self.estimate_distance(
                xyz_points
            )

            position = self.estimate_object_position(
                xyz_points
            )

            dimensions = self.estimate_object_dimensions(
                xyz_points
            )

            yaw = self.estimate_object_yaw(
                xyz_points
            )

            if position is None:
                continue

            world_objects.append(
                {
                    "class": obj["class_name"],
                    "camera": camera_name,

                    # BEV visualization
                    "position": position,
                    "distance": distance,

                    # Future 3D MOT
                    "center": position,
                    "score": obj["score"],
                    "dimensions": dimensions,
                    "yaw": yaw,
                    
                    # Optional
                    "cloud": obj["cloud"]
                }
            )

        return image, world_objects

    def draw_object_cloud(
        self,
        image,
        object_cloud
    ):

        for point in object_cloud:

            x, y = point["pixel"]

            cv2.circle(
                image,
                (x, y),
                2,
                (0, 255, 0),
                -1
            )

        return image

    def estimate_distance(
        self,
        xyz_points
    ):

        if len(xyz_points) == 0:

            return None

        distances = np.linalg.norm(
            xyz_points,
            axis=1
        )

        distance = np.percentile(
            distances,
            10
        )

        return float(distance)


    def estimate_object_dimensions(
        self,
        xyz_points
    ):
        """
        Estimate object dimensions from the LiDAR point cloud.

        Returns
        -------
        tuple
            (length, width, height)
        """

        # xyz_points = self.process_object_clouds(
        #     object_cloud
        # )

        if len(xyz_points) == 0:
            return None

        min_xyz = np.min(
            xyz_points,
            axis=0
        )

        max_xyz = np.max(
            xyz_points,
            axis=0
        )

        length = float(max_xyz[0] - min_xyz[0])
        width = float(max_xyz[1] - min_xyz[1])
        height = float(max_xyz[2] - min_xyz[2])

        return (
            length,
            width,
            height
        )

    def estimate_object_yaw(
        self,
        xyz_points
    ):
        """
        Estimate object yaw (heading) from the LiDAR point cloud.

        Returns
        -------
        float
            Yaw in radians (ego frame).
        """

        # xyz_points = self.process_object_clouds(
        #     object_cloud
        # )

        if len(xyz_points) < 2:
            return 0.0

        #
        # Use ground-plane coordinates only
        #

        xy = xyz_points[:, :2]

        #
        # Center the points
        #

        xy = xy - np.mean(
            xy,
            axis=0
        )

        #
        # PCA
        #

        covariance = np.cov(
            xy,
            rowvar=False
        )

        eigenvalues, eigenvectors = np.linalg.eigh(
            covariance
        )

        #
        # Principal axis
        #

        principal_axis = eigenvectors[
            :,
            np.argmax(eigenvalues)
        ]

        #
        # Heading
        #

        yaw = np.arctan2(
            principal_axis[1],
            principal_axis[0]
        )

        return float(yaw)


    def process_object_clouds(
        self,
        object_cloud
    ):

        if len(object_cloud) == 0:

            return np.empty(
                (0, 3),
                dtype=np.float32
            )

        xyz = np.array(
            [
                point["xyz"]
                for point in object_cloud
            ],
            dtype=np.float32
        )

        return xyz

    def estimate_object_position(
        self,
        xyz_points
    ):
        """
        Estimate object position using the
        front surface of the LiDAR cluster.
        """

        # xyz_points = self.process_object_clouds(
        #     obj_cloud
        # )

        if len(xyz_points) == 0:
            return None

        #
        # Distance of each LiDAR point
        #

        distances = np.linalg.norm(
            xyz_points,
            axis=1
        )

        #
        # Front surface threshold
        #

        threshold = np.percentile(
            distances,
            10
        )

        #
        # Keep only the front-most points
        #

        front_points = xyz_points[
            distances <= threshold
        ]

        #
        # Fallback
        #

        if len(front_points) == 0:
            front_points = xyz_points

        #
        # Representative position
        #

        position = np.mean(
            front_points,
            axis=0
        )

        return position



    def process_camera(
        self,
        image,
        lidar,
        result,
        camera_name
    ):
        """
        project_lidar()
            ↓
        extract_object_clouds()
            ↓
        process_object_clouds_and_distance()
        """
        pass

File: perception_cpp/src/perception_3d_pipeline.cpp
#include "perception_3d_pipeline.hpp"

#include <cmath>
#include <stdexcept>
#include <algorithm>

Perception3DPipeline::Perception3DPipeline()
    : fx(320.0f),
      fy(320.0f),
      cx(320.0f),
      cy(240.0f)
{
    camera_yaw["front"] = 0.0f;
    camera_yaw["rear"]  = 180.0f;
    camera_yaw["left"]  = -90.0f;
    camera_yaw["right"] = 90.0f;
}

Perception3DPipeline::ProjectionResult
Perception3DPipeline::project_lidar(
    const std::vector<Eigen::Vector3f>& lidar,
    const std::string& camera_name,
    int image_width,
    int image_height
)
{
    ProjectionResult result;    
    

    // Camera configuration
    float yaw_deg = camera_yaw.at(camera_name);

    float yaw =
        -yaw_deg * static_cast<float>(M_PI) / 180.0f;

    // Rotation matrix
    Eigen::Matrix3f R;

    R << std::cos(yaw), -std::sin(yaw), 0.0f,
         std::sin(yaw),  std::cos(yaw), 0.0f,
         0.0f,           0.0f,          1.0f;

    // Process each LiDAR point
    for (const auto& ego_point : lidar)
    {
        // LiDAR -> camera frame
        Eigen::Vector3f point = R * ego_point;

        // LiDAR -> camera translation
        point.x() -= 1.5f;

        // CARLA axes -> camera axes
        float X = point.y();
        float Y = -point.z();
        float Z = point.x();

        // Keep points in front
        if (Z <= 0.0f)
            continue;

        // Project to image
        float u = fx * X / Z + cx;
        float v = fy * Y / Z + cy;

        // Keep points inside image
        if (u < 0.0f ||
            u >= static_cast<float>(image_width) ||
            v < 0.0f ||
            v >= static_cast<float>(image_height))
        {
            continue;
        }

        result.u.push_back(static_cast<int>(u));
        result.v.push_back(static_cast<int>(v));

        // Keep original ego-frame point
        result.ego_points.push_back(ego_point);
    }

    return result;
}
std::vector<Perception3DPipeline::ObjectCloud>
Perception3DPipeline::extract_object_clouds(
    const std::vector<std::vector<std::vector<float>>>& masks,
    const std::vector<Eigen::Vector4f>& boxes,
    const std::vector<int>& classes,
    const std::vector<std::string>& class_names,
    const std::vector<int>& u,
    const std::vector<int>& v,
    const std::vector<Eigen::Vector3f>& projected_points,
    int image_width,
    int image_height
)
{
    std::vector<ObjectCloud> object_clouds;
    object_clouds.reserve(masks.size());

    for (size_t object_id = 0; object_id < masks.size(); ++object_id)
    {
        const auto& mask_data = masks[object_id];

        if (mask_data.empty() || mask_data[0].empty())
        {
            continue;
        }

        int mask_height = static_cast<int>(mask_data.size());
        int mask_width = static_cast<int>(mask_data[0].size());

        // Build uint8 mask directly, row by row, via raw pointers
        // (skips the CV_32FC1 intermediate + convertTo pass)
        cv::Mat mask_uint8(mask_height, mask_width, CV_8UC1);

        for (int y = 0; y < mask_height; ++y)
        {
            const float* src = mask_data[y].data();
            uint8_t* dst = mask_uint8.ptr<uint8_t>(y);

            for (int x = 0; x < mask_width; ++x)
            {
                dst[x] = static_cast<uint8_t>(src[x]);
            }
        }

        cv::Mat resized_mask;

        cv::resize(
            mask_uint8,
            resized_mask,
            cv::Size(image_width, image_height),
            0.0,
            0.0,
            cv::INTER_NEAREST
        );

        // Class
        int class_id = classes[object_id];
        std::string class_name = "unknown";

        if (class_id >= 0 &&
            class_id < static_cast<int>(class_names.size()))
        {
            class_name = class_names[class_id];
        }

        // Create object cloud
        ObjectCloud object;
        object.cloud.reserve(u.size());

        for (size_t i = 0; i < u.size(); ++i)
        {
            int x = u[i];
            int y = v[i];

            if (x < 0 || x >= image_width ||
                y < 0 || y >= image_height)
            {
                continue;
            }

            if (resized_mask.ptr<uint8_t>(y)[x] > 0)
            {
                ObjectPoint point;
                point.u = x;
                point.v = y;
                point.xyz = projected_points[i];

                object.cloud.push_back(std::move(point));
            }
        }

        object.box = boxes[object_id];
        object.class_name = class_name;

        object_clouds.push_back(std::move(object));
    }

    return object_clouds;
}

Perception3DPipeline::ProcessedObjectsResult
Perception3DPipeline::process_object_clouds_and_distance(
    cv::Mat& image,
    const std::vector<ObjectCloud>& object_clouds,
    const std::string& camera_name
)
{
    ProcessedObjectsResult result;

    result.image = image.clone();

    for (const auto& obj : object_clouds)
    {
        // --------------------------------------------------
        // Draw object cloud
        // --------------------------------------------------

        for (const auto& point : obj.cloud)
        {
            cv::circle(
                result.image,
                cv::Point(point.u, point.v),
                2,
                cv::Scalar(0, 255, 0),
                -1
            );
        }


        // --------------------------------------------------
        // No LiDAR points
        // --------------------------------------------------

        if (obj.cloud.empty())
        {
            continue;
        }


        // --------------------------------------------------
        // Calculate distances
        // --------------------------------------------------

        std::vector<float> distances;

        distances.reserve(obj.cloud.size());

        for (const auto& point : obj.cloud)
        {
            distances.push_back(point.xyz.norm());
        }


        // --------------------------------------------------
        // 10th percentile
        // --------------------------------------------------

        std::vector<float> sorted_distances = distances;

        std::sort(
            sorted_distances.begin(),
            sorted_distances.end()
        );

        size_t percentile_index =
            static_cast<size_t>(
                0.10f * sorted_distances.size()
            );

        if (percentile_index >= sorted_distances.size())
        {
            percentile_index =
                sorted_distances.size() - 1;
        }

        float threshold =
            sorted_distances[percentile_index];

        float distance = threshold;


        // --------------------------------------------------
        // Front-most points
        // --------------------------------------------------

        std::vector<Eigen::Vector3f> front_points;

        for (size_t i = 0; i < obj.cloud.size(); ++i)
        {
            if (distances[i] <= threshold)
            {
                front_points.push_back(
                    obj.cloud[i].xyz
                );
            }
        }


        // --------------------------------------------------
        // Fallback
        // --------------------------------------------------

        if (front_points.empty())
        {
            for (const auto& point : obj.cloud)
            {
                front_points.push_back(point.xyz);
            }
        }


        // --------------------------------------------------
        // Mean position
        // --------------------------------------------------

        Eigen::Vector3f position =
            Eigen::Vector3f::Zero();

        for (const auto& point : front_points)
        {
            position += point;
        }

        position /= static_cast<float>(
            front_points.size()
        );


        // --------------------------------------------------
        // Create world object
        // --------------------------------------------------

        WorldObject world_object;

        world_object.class_name = obj.class_name;
        world_object.camera = camera_name;
        world_object.box = obj.box;
        world_object.position = position;
        world_object.distance = distance;

        result.world_objects.push_back(
            world_object
        );
    }

    return result;
}